src/common.rs

```diff
@@ -1,52 +1,133 @@
-    validators::RpcValidatorRequest,
-    EXPERIMENTAL_genesis_config::{RpcGenesisConfigRequest, RpcGenesisConfigResponse},
-};
-use near_primitives::views::EpochValidatorInfo;
-
-fn validators_info(
-    epoch: near_primitives::types::EpochReference,
-    network_config: &near_cli_rs::config::NetworkConfig,
-) -> (RpcGenesisConfigResponse, EpochValidatorInfo) {
-    let client = network_config.json_rpc_client();
-    let genesis_config = client.blocking_call(RpcGenesisConfigRequest).unwrap();
-    let validator_info = client
-        .blocking_call(&RpcValidatorRequest {
-            epoch_reference: epoch,
+use color_eyre::eyre::ContextCompat;
+use num_rational::Rational32;
+
+use near_primitives::views::{CurrentEpochValidatorInfo, NextEpochValidatorInfo};
+
+pub fn find_seat_price(
+    validators: Vec<CurrentOrNextValidatorInfoOrProposalsTable>,
+    max_number_of_seats: u64,
+    minimum_stake_ratio: Rational32,
+    protocol_version: near_primitives::types::ProtocolVersion,
+) -> color_eyre::eyre::Result<near_cli_rs::common::NearBalance> {
+    if protocol_version < 49 {
+        return find_seat_price_for_protocol_before_49(validators, max_number_of_seats);
+    }
+    find_seat_price_for_protocol_after_49(
+        validators,
+        max_number_of_seats,
+        vec![*minimum_stake_ratio.numer(), *minimum_stake_ratio.denom()],
+    )
+}
+
+fn find_seat_price_for_protocol_before_49(
+    validators: Vec<CurrentOrNextValidatorInfoOrProposalsTable>,
+    num_seats: u64,
+) -> color_eyre::eyre::Result<near_cli_rs::common::NearBalance> {
+    let stakes = validators
+        .iter()
+        .map(|validator_info| match validator_info {
+            CurrentOrNextValidatorInfoOrProposalsTable::CurrentEpochValidatorInfo(current) => {
+                current.stake
+            }
+            CurrentOrNextValidatorInfoOrProposalsTable::NextEpochValidatorInfo(next) => next.stake,
+            CurrentOrNextValidatorInfoOrProposalsTable::ProposalsTable(proposals_table) => {
+                proposals_table.stake
+            }
+        })
+        .collect::<Vec<_>>();
+    let stakes_sum: u128 = stakes.iter().sum();
+    if stakes_sum < num_seats as u128 {
+        return Err(color_eyre::eyre::Report::msg("Stakes are below seats"));
+    }
+    let mut left: u128 = 1;
+    let mut right: u128 = stakes_sum + 1;
+    while left != (right - 1) {
+        let mid = left.saturating_add(right) / 2;
+        let mut found = false;
+        let mut current_sum: u128 = 0;
+        for stake in stakes.clone() {
+            current_sum = current_sum.saturating_add(stake.saturating_div(mid));
+            if current_sum >= num_seats as u128 {
+                left = mid;
+                found = true;
+                break;
+            }
+        }
+        if !found {
+            right = mid;
+        }
+    }
+    Ok(near_cli_rs::common::NearBalance::from_yoctonear(left))
+}
+
+fn find_seat_price_for_protocol_after_49(
```

Comment:
Add a link to the relevant code in near-api-js where you took the implementation from:

```suggestion
/// This implementation is ported from near-api-js:
/// https://github.com/near/near-api-js/blob/bdbf839f54fbc399d7299da8cf9966bbdc426238/packages/utils/src/validators.ts#L52-L64
fn find_seat_price_for_protocol_after_49(
```